trieste/acquisition/function.py

```diff
@@ -745,9 +793,20 @@ def prepare_acquisition_function(
         feasible_query_points = tf.boolean_mask(objective_dataset.query_points, is_feasible)
         feasible_mean, _ = objective_model.predict(feasible_query_points)
         eta = tf.reduce_min(feasible_mean, axis=0)
-        function = expected_improvement(objective_model, eta)
+        tf.debugging.Assert(isinstance(self._improvement_fn, expected_improvement), [])
+        self._improvement_fn.update(eta)  # type: ignore
+
+        if self._constrained_fn is not None:
+            return self._constrained_fn
+
+        @tf.function
+        def constrained_function(x: TensorType) -> TensorType:
+            return cast(AcquisitionFunction, self._improvement_fn)(x) * cast(
+                AcquisitionFunction, self._constraint_fn
+            )(x)
 
-        return lambda at: function(at) * constraint_fn(at)
+        self._constrained_fn = constrained_function
```

Comment:
There is a cast in similar situation in `prepare`

Reply:
See comment above. The reason a cast isn't necessary here is that mypy infers that self._constrained_fn can't be None due to the earlier if clause. (Though as I said, I'll remove the cast from further up too.)